go/enclave/db/storage.go

```diff
@@ -372,3 +471,119 @@ func (s *storageImpl) StoreRollup(rollup *core.Rollup) error {
+	query := "select distinct topic0, topic1, topic2, topic3, topic4, datablob, blockHash, blockNumber, txHash, txIdx, logIdx, address from events where 1=1 "
+	var queryParams []any
+
+	// todo - once we introduce streaming, this check should be converted into an assert. This should never return non-account specific logs
+	if requestingAccount != nil {
+		// Add relevancy rules
+		//  An event is considered relevant to all account owners whose addresses are used as topics in the event.
+		//	In case there are no account addresses in an event's topics, then the event is considered relevant to everyone (known as a "lifecycle event").
+		query += " AND (lifecycleEvent OR (relAddress1=? OR relAddress2=? OR relAddress3=? OR relAddress4=?)) "
+		queryParams = append(queryParams, requestingAccount.Bytes())
+		queryParams = append(queryParams, requestingAccount.Bytes())
+		queryParams = append(queryParams, requestingAccount.Bytes())
+		queryParams = append(queryParams, requestingAccount.Bytes())
+	}
+
+	query += whereCondition
+	queryParams = append(queryParams, whereParams...)
+
+	// fmt.Printf("%s , %v\n", query, queryParams)
+	rows, err := s.db.GetSQLDB().Query(query, queryParams...)
+	if err != nil {
+		return nil, err
+	}
+	for rows.Next() {
+		l := types.Log{
+			Topics: []gethcommon.Hash{},
+		}
+		var t0, t1, t2, t3, t4 sql2.NullString
+		err := rows.Scan(&t0, &t1, &t2, &t3, &t4, &l.Data, &l.BlockHash, &l.BlockNumber, &l.TxHash, &l.TxIndex, &l.Index, &l.Address)
+		if err != nil {
+			return nil, fmt.Errorf("could not load log entry from db: %w", err)
+		}
+		if t0.Valid {
+			l.Topics = append(l.Topics, hash(t0))
+		}
+		if t1.Valid {
+			l.Topics = append(l.Topics, hash(t1))
+		}
+		if t2.Valid {
+			l.Topics = append(l.Topics, hash(t2))
+		}
+		if t3.Valid {
+			l.Topics = append(l.Topics, hash(t3))
+		}
+		if t4.Valid {
+			l.Topics = append(l.Topics, hash(t4))
+		}
+
+		result = append(result, &l)
+	}
+	err = rows.Close()
+	if err != nil {
+		return result, err
+	}
+
+	return result, nil
+}
+
+func (s *storageImpl) FilterLogs(requestingAccount *gethcommon.Address, filter *filters.FilterCriteria) ([]*types.Log, error) {
+	// transform a FilterCriteria into an SQL where condition
+	queryParams := []any{}
+	query := ""
+	if filter.BlockHash != nil {
+		query += " AND blockHash = ?"
+		queryParams = append(queryParams, filter.BlockHash.Bytes())
+	}
+	if filter.FromBlock != nil {
+		query += " AND blockNumber >= ?"
+		queryParams = append(queryParams, filter.FromBlock.Int64())
+	}
+	if filter.ToBlock != nil {
+		query += " AND blockNumber < ?"
+		queryParams = append(queryParams, filter.ToBlock.Int64())
+	}
+	if len(filter.Addresses) > 0 {
+		query += " AND address in (?" + strings.Repeat(",?", len(filter.Addresses)-1) + ")"
+		for _, address := range filter.Addresses {
+			queryParams = append(queryParams, address.Bytes())
+		}
+	}
+	if len(filter.Topics) > 4 {
+		return nil, fmt.Errorf("invalid filter. Too many topics")
```

Comment:
Have you got an off by one error here? Have seen you reference topics t0-t4 in a few places. Or is it that filters only have max 4 topics because one of those topics is not filterable?